File: algo/98. Validate Binary Search Tree.js
// @9 Divide and Conquer

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */

var helper = function(root) {
    
    if (root == null) {
        return undefined;
    }
    
    // true or false, max, min
    if (root.left == null && root.right == null) {
        return [true, root.val, root.val];
    }
    
    var leftRes = helper(root.left);
    var rightRes = helper(root.right);
    
    if (root.left == null) {
        if (root.val < rightRes[2] && rightRes[0] == true) {
            return [true, rightRes[1], root.val];
        }
    } else if (root.right == null) {
        if (root.val > leftRes[1] && leftRes[0] == true) {
            return [true, root.val, leftRes[2]];
        }
    } else {
        if (root.val < rightRes[2] && root.val > leftRes[1] && rightRes[0] == true && rightRes[0] == true) {
            return [true, rightRes[1], leftRes[2]];
        }
    }

    return [false];
}

var isValidBST = function(root) {
    if (root==null) {
        return true;
    }
    return helper(root)[0];
};

File: round3/98. Validate Binary Search Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool IsValidBST(TreeNode root) {
        return Helper(root, long.MinValue, long.MaxValue);
    }
    
    public bool Helper(TreeNode root, long min, long max) {
        if (root == null) return true;
        if (root.val >= max || root.val <= min) return false;
        
        bool leftRes = Helper(root.left, min, root.val);
        bool rightRes = Helper(root.right, root.val, max);
        
        if (leftRes && rightRes) return true;
        return false;
    }
}